ARROW-8063: [Python][Dataset] Start user guide for pyarrow.dataset (#6779)
apache/arrow: docs/source/python/dataset.rst

.. distributed with this work for additional information
.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

.. currentmodule:: pyarrow.dataset

.. _dataset:

Tabular Datasets
================

.. warning::

    The ``pyarrow.dataset`` module is experimental (specifically the classes),
    and a stable API is not yet guaranteed.

The ``pyarrow.dataset`` module provides functionality to efficiently work with
tabular, potentially larger than memory and multi-file, datasets:

* A unified interface for different sources: supporting different sources
  (files, database connection, ..), different file systems (local, cloud) and

>   different file formats (Parquet, CSV, JSON, Feather, ..)

Author: Yeah, if we keep it, I should at least make it clearer this is "the vision" and not reality ..

If we save the table as a Feather file instead of Parquet files:

.. ipython:: python

    import pyarrow.feather as feather

    feather.write_feather(table, base / "data.feather")

then we can read the Feather file using the same functions, but with specifying
``format="ipc"``:

.. ipython:: python

    dataset = ds.dataset(base / "data.feather", format="ipc")
    dataset.to_table().to_pandas().head()

The format name as a string, like:

.. code-block::

    ds.dataset(..., format="parquet")

is a short-hand for

.. code-block::

    ds.dataset(..., format=ds.ParquetFileFormat())

Those file format objects can take keywords to customize reading (see for
example :class:`ParquetFileFormat`).

Filtering data
--------------

To avoid reading all data when only needing a subset, the ``columns`` and
``filter`` keywords can be used.

The ``columns`` keyword can be used to only read a selection of the columns,
and accepts a list of column names:

.. ipython:: python

    dataset = ds.dataset(base / "parquet_dataset", format="parquet")
    dataset.to_table(columns=['a', 'b']).to_pandas()

With the ``filter`` keyword, rows which do not match the filter predicate will

> not be included in the returned table. The keyword expects a boolean expression

Reviewer: ```suggestion
not be included in the returned table. The keyword expects a class:`Expression`
```

Author: You find the "boolean" wrong here?

Reviewer: It's not wrong, just less specific than referencing the class `Expression` explicitly. For example a reader might imagine that `filter=True` or even `filter=lambda a: a % 2 == 0` were legal

Author: OK, understand now, made it "boolean :class:'Expression'" then

> involving one of the columns, and those expressions can be created using the
> :func:`field` helper function:

Reviewer: ```suggestion
constraining at least one of the columns::
```

Author: For non-native speakers (as myself), I think "constraining" might be difficult to understand correctly here.

Reviewer: fair enough, how about "referencing" instead?

Author: Ah, I like that :-)

.. ipython:: python

    dataset.to_table(filter=ds.field('a') >= 7).to_pandas()
    dataset.to_table(filter=ds.field('c') == 2).to_pandas()

> Any column can be referenced using the :func:`field` function (which creates a
> :class:`FieldExpression`), and many different expressions can be created,
> including the standard boolean comparison operators (equal, larger/less than,
> etc), but also set membership testing:

Reviewer: ```suggestion
Any column - not just partition columns - can be referenced using the
:func:`field` function (which creates a :class:`FieldExpression`). Operator
overloads are provided to compose filters including the comparisons
(equal, larger/less than, etc), set membership testing, and boolean
combinations (and, or, not):
```

.. ipython:: python

    ds.field('a') != 3

>     ds.field('a').isin([1, 2, 3])

Reviewer: ```suggestion
    ds.field('a').isin([1, 2, 3])
    ds.field('a') > ds.field('b') & ds.field('b') > 1
```

Reading partitioned data
------------------------

> A dataset consisting of a flat directory with files was already shown above.

Reviewer: ```suggestion
Above, a dataset consisting of a flat directory with files was shown.
```

> But the dataset can also contain nested directories defining a partitioned

Reviewer: ```suggestion
However a dataset can exploit a nested directory structure defining a partitioned
```

in Apache Hive.

Let's create a small partitioned dataset. The:func:`~pyarrow.parquet.write_to_dataset`
function can write such hive-like partitioned datasets.

.. ipython:: python

    table = pa.table({'a': range(10), 'b': np.random.randn(10), 'c': [1, 2] * 5,
                      'part': ['a'] * 5 + ['b'] * 5})
    pq.write_to_dataset(table, str(base / "parquet_dataset_partitioned"),
                        partition_cols=['part'])

The above created a directory with two subdirectories ("part=a" and "part=b"),
and the Parquet files written in those directories no longer include the "part"
column.

Reading this dataset with :func:`dataset`, we now specify that the dataset
uses a hive-like partitioning scheme with the `partitioning` keyword:

.. ipython:: python

    dataset = ds.dataset(str(base / "parquet_dataset_partitioned2"), format="parquet",
                         partitioning="hive")
    dataset.files

Although the partition fields are not included in the actual Parquet files,
they will be added back to the resulting table when scanning this dataset:

.. ipython:: python

    dataset.to_table().to_pandas().head(3)

> We can now filter on the partition keys, which avoids loading certain files
> at all if they do not match the predicate:

Reviewer: ```suggestion
We can now filter on the partition keys, which avoids loading files
altogether if they do not match the predicate:
```

.. ipython:: python

    dataset.to_table(filter=ds.field("part") == "b").to_pandas()

Different partitioning schemes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The above example uses a hive-like directory scheme, such as "/year=2009/month=11/day=15".
We specified this passing the ``partitioning="hive"`` keyword. In this case,
the types of the partition keys are inferred from the file paths.

It is also possible to explicitly define the schema of the partition keys
using the :func:`partitioning` function. For example:

.. code-block::

    part = ds.partitioning(
        pa.schema([("year", pa.int16()), ("month", pa.int8()), ("day", pa.int32())]),
        flavor="hive"
    )
    dataset = ds.dataset(..., partitioning=part)

> In addition to the hive-like directory scheme, also a "directory partitioning"
> scheme is supported, where the segments in the file path are the values of
> the partition keys without including the name. The equivalent (year, month, day)
> example would be "/2019/11/15/".

Reviewer: ```suggestion
"Directory partitioning" is also supported, where the segments in the file path represent the values of
partition keys with the field name implicit in the segment's index. For example, given field names
"year", "month", and "day", one path might be "/2019/11/15".
```

> Since the names are not included in the file paths, those need to be specified:

Reviewer: ```suggestion
Since the names are not included in the file paths, these must be specified
when constructing a directory partitioning:
```

.. code-block::

    part = ds.partitioning(field_names=["year", "month", "day"])

> Also here, a full schema can be provided to not let the types be inferred
> from the file paths.

Reviewer: ```suggestion
Directory directory partitioning also supports providing a full schema rather than inferring
types from file paths.
```

Reading from cloud storage
--------------------------

- example with s3 filesystem / hdfs filesystem

Manual specification of the Dataset
-----------------------------------

> The :func:`dataset` function allows to easily create a Dataset from a directory,
> crawling all subdirectories for files and partitioning information. However,
> when the dataset files and partitions are already known (for example, when this
> information is stored in metadata), it is also possible to create a Dataset
> explicitly using the lower level API without any automatic discovery or
> inference.

Reviewer: ```suggestion
The :func:`dataset` function allows easy creation of a Dataset viewing a directory,
crawling all subdirectories for files and partitioning information. However
sometimes discovery is not required and the dataset's files and partitions
are already known (for example, when this information is stored in metadata).
In this case it is possible to create a Dataset explicitly without any
automatic discovery or inference.
```

For the example here, we are going to use a dataset where the file names contain
additional partitioning information:

.. ipython:: python

>     # creating a dummy dataset: directory with two files
>     table = pa.table({'col1': range(3), 'col2': np.random.randn(3)})
>     pq.write_table(table, "parquet_dataset_manual/data_file1.parquet")
>     pq.write_table(table, "parquet_dataset_manual/data_file2.parquet")
> 
> To create a Dataset from a list of files, we need to specify the schema, format,
> filesystem, and paths manually:
> 
> .. ipython:: python
> 
>     import pyarrow.fs
> 
>     schema = pa.schema([("file", pa.int64()), ("col1", pa.int64()), ("col2", pa.float64())])
> 
>     dataset = ds.FileSystemDataset(
>         schema, None, ds.ParquetFileFormat(), pa.fs.LocalFileSystem(),
>         ["parquet_dataset_manual/data_file1.parquet", "parquet_dataset_manual/data_file2.parquet"],
>         [ds.field('file') == 1, ds.field('file') == 2])

Reviewer: ```suggestion
    # creating a dummy dataset: directory with two files
    table = pa.table({'col1': range(3), 'col2': np.random.randn(3)})
    pq.write_table(table, "parquet_dataset_manual/old.parquet")
    pq.write_table(table, "parquet_dataset_manual/new.parquet")

To create a Dataset from a list of files, we need to specify the schema, format,
filesystem, and paths manually:

.. ipython:: python

    import pyarrow.fs

    schema = pa.schema([("year", pa.int32()), ("col1", pa.int64()), ("col2", pa.float64())])

    dataset = ds.FileSystemDataset(
        schema, None, ds.ParquetFileFormat(), pa.fs.LocalFileSystem(),
        ["parquet_dataset_manual/old.parquet", "parquet_dataset_manual/new.parquet"],
        [ds.field('year') < 2020, ds.field('year') >= 2020])
```

Author: I like the idea of your improvement, but this is actually an interesting case. If you do such a larger/smaller filter, it can't really be added to the table as a column (just tried, and currently it gives all nulls). 

While interesting, it might not be the best example to show first.

But will change to using years in the file names (but just with exact names), that will already make the example nicer.

Author: And also, filtering with that kind of expressions is also not very intuitively. 
If you have the partition expressions as `[ds.field('year') < 2020, ds.field('year') >= 2020])`, then doing a filter with `ds.field('year') == 2020`, you might intuitively expect that it will read the the second file, as "year == 2020" is a subset of "year >= 2020", but this is not the case.

Reviewer: > you might intuitively expect that it will read the the second file, as "year == 2020" is a subset of "year >= 2020", but this is not the case.

I certainly did expect that. I will investigate.

Author: Also the change here from int64 -> int32 in the schema doesn't work, because the partition expressions specified for the files are default int64 ..

> We also specified the "partition expressions" for our files, so this information
> is included when reading the data and can be used for filtering:

Reviewer: ```suggestion
Since we specified the "partition expressions" for our files, this information
is materialized as columns when reading the data and can be used for filtering:
```

.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

.. currentmodule:: pyarrow.dataset

.. _dataset:

Tabular Datasets
================

.. warning::

    The ``pyarrow.dataset`` module is experimental (specifically the classes),
    and a stable API is not yet guaranteed.

The ``pyarrow.dataset`` module provides functionality to efficiently work with
tabular, potentially larger than memory and multi-file, datasets:

* A unified interface for different sources: supporting different sources
  (files, database connection, ..), different file systems (local, cloud) and
  different file formats (Parquet, CSV, JSON, Feather, ..)
* Discovery of sources (crawling directories, handle directory-based partitioned
  datasets, basic schema normalization, ..)

> * Optimized reading with pedicate pushdown (filtering rows), projection

Reviewer 2: "predicate"

  (selecting columns), parallel reading or fine-grained managing of tasks.

For those familiar with the existing :class:`pyarrow.parquet.ParquetDataset` for
reading Parquet datasets: ``pyarrow.dataset``'s goal is similar but not specific
to the Parquet format and not tied to Python: the same datasets API is exposed
in the R bindings or Arrow. In addition ``pyarrow.dataset`` boasts improved
perfomance and new features (e.g. filtering within files rather than only on
partition keys).

Reading Datasets
----------------

.. TODO Full blown example with NYC taxi data to show off, afterwards explain all parts:

> .. ipython:: python

Reviewer 2: Hmm, I'm frankly not fond of "ipython" directives. The more we add, the slower building the docs becomes and the more it deters from editing/improving them.

Author: Yeah, I am happy to change this (I mostly used it because it was already used elsewhere). 

However, I think ideally we still check the code examples for correctness, where applicable. For example by using pytest doctests on them. This can be run separate as tests, and doesn't need to be part of doc building.

Manual scheduling
-----------------

..
    Possible content:
    - fragments (get_fragments)
    - scan / scan tasks / iterators of record batches

The :func:`~Dataset.to_table` method loads all selected data into memory

> at once resulting in a pyarrow Table. Alternatively, a datasetscan also be

Reviewer: ```suggestion
at once resulting in a pyarrow Table. Alternatively, a dataset also be
```
